include/SDL_pen.h

```diff
@@ -0,0 +1,316 @@
+ *          etc.) or calls the event or windowing subsystems, due to
+ *          hot-plugging.
+ *
+ * \since This function is available since SDL 2.TBD
+ *
+ * \sa SDL_PenIDForIndex()
+ */
+extern DECLSPEC int SDLCALL SDL_NumPens(void);
+
+/**
+ * Retrieves a pen attached to the system while iterating over all pens.
+ *
+ * This function is intended for enumerating all currently attached pens.
+ * The iteration order and number of valid indices may change after
+ * any call to event processing (e.g., ::SDL_PollEvent()) or to the graphics, or windowing subsystems.
+ *
+ * Use ::SDL_PenID to track pens throughout a session, or ::SDL_PenGUID for
+ * tracking across sessions.
+ *
+ * \param device_index An index between 0 and ::SDL_NumPens() - 1 (inclusive).
+ *
+ * \returns An ::SDL_PenID.  ::SDL_PENID_VALID holds iff the "device_index" was in the valid range,
+ *     otherwise ::SDL_GetError() is set.
+ *
+ * \since This function is available since SDL 2.TBD
+ *
+ * \sa SDL_NumPens()
+ */
+extern DECLSPEC SDL_PenID SDLCALL SDL_PenIDForIndex(int device_index);
+
+/**
+ * Retrieves the pen's current status.
+ *
+ * If the pen is detached (cf. ::SDL_PenAttached), this operation may return
+ * default values.
+ *
+ * \param pen The pen to query.
+ * \param[out] x Out-mode parameter for pen x coordinate.  May be NULL.
+ * \param[out] y Out-mode parameter for pen y coordinate.  May be NULL.
+ * \param[out] axes Out-mode parameter for axis information.  May be null.  The axes are in the same order as for
+ *     \link SDL_PEN_AXES \endlink.
+ * \param num_axes Maximum number of axes to write to "axes".
+ *
+ * \returns a bit mask with the current pen button states (::SDL_BUTTON_LMASK etc.) and exactly one of
+ *     ::SDL_PEN_INK_MASK or ::SDL_PEN_ERASER_MASK , or 0 on error (see ::SDL_GetError()).
+ *
+ * \since This function is available since SDL 2.TBD
+ */
+extern DECLSPEC Uint32 SDLCALL SDL_PenStatus(SDL_PenID pen, float * x, float * y, float * axes, size_t num_axes);
+
+/**
+ * Retrieves an ::SDL_PenID for the given ::SDL_PenGUID.
+ *
+ * \param guid A pen GUID.
+ *
+ * \returns An ::SDL_PenID.  ::SDL_PENID_VALID holds iff the "device_index" was in the valid range,
+ *     otherwise ::SDL_GetError() is set.
+ *
+ * \since This function is available since SDL 2.TBD
+ *
+ * \sa SDL_PenGUID(), SDL_PenGUIDForString()
+ */
+extern DECLSPEC SDL_PenID SDLCALL SDL_PenIDForGUID(SDL_PenGUID guid);
+
+/**
+ * Retrieves the ::SDL_PenGUID for a given ::SDL_PenID.
+ *
+ * \param penid The pen to query.
+ *
+ * \returns The corresponding pen GUID; persistent across multiple sessions.
+ *     If "penid" is not ::SDL_PENID_VALID(), returns an all-zeroes GUID.
+ *
+ * \since This function is available since SDL 2.TBD
+ *
+ * \sa SDL_PenForID(), SDL_PenStringForGUID(), SDL_PenGUIDCompare()
+ */
+extern DECLSPEC SDL_PenGUID SDLCALL SDL_PenGUIDForPenID(SDL_PenID penid);
+
+/**
+ * Compares two ::SDL_PenGUID objects to determine their order.
+ *
```

Comment:
What does it mean for a pen to be left or right? What does the pen ordering signify?

Reply:
Changing to:

"Compares two ::SDL_PenGUID objects to determine the lexicographic order of their bit strings.

"The order imposed by this function is arbitrary but persistent across sessions,
 if the hardware reports persistent ::SDL_PenGUID objects."

Reply:
Why is this function useful at the public API level?

Reply:
Both joystick and pen GUIDs are intended as keys (in the database/map sense), so clients want to compare them when handling config data.
GUIDs in SDL aren't *really* opaque/abstract, since their internals are public, but clients could get away with thinking so, since the to-string/from-string operations give them enough to do comparisons with. This is clearly clunky, but the safer/simpler way requires looking up/understanding what GUIDs are and using `memcmp()`.

The tradeoff I proposed here was to add a public API function so that clients that use GUIDs don't have to look up those internals when comparing/ordering GUIDs.

An alternative would be to define a macro instead-- no symbol, no overhead for people who don't use this, no binary compatibility debt.  Errors would be weirder, though.

Given the above reasoning, should I (a) keep the operation, (b) remove it, (c) define a macro instead?

Reply:
NB: This is connected to the the `SDL_guid.c` comments below.

Reply:
Removed as part of that discussion.